### DISCUSSION TITLE
HFSS Analysis progress in non-graphical mode

### DISCUSSION THREAD

**Discussion by nathmay** [ansys/pyaedt#4774] [Q&A] [answered]
When HFSS is launched in non-graphical mode, and `analyze()` is executed, is there a way to see or retrieve the progress? With `blocking=False`, I was wondering if there is a way to monitor the solution progress through script.

```python
from pyaedt import Hfss

with Hfss(non_graphical=True, close_on_exit=True) as aedtapp:
    # Solve
    aedtapp.analyze(blocking=False)
    
    # Monitor solution progress here, while performing other tasks

    # Save
    aedtapp.save_project()
```

**Comment by Samuelopez-ansys** [comment]
Hi @nathmay ,

If blocking=False as you know, the script will not stop. You can get the profile info:

desktop.get_monitor_data

or just check if it is running:

are_there_simulations_running

You can also stop the simulation:

stop_simulations

**Reply by nathmay** [reply]
Hi @Samuelopez-ansys, thank you for your answer! I tried using `desktop.get_monitor_data()`, but couldn't get it working somehow. I get the following error trace:
```
PyAEDT ERROR:   File "C:\Users\nathm\test.py", line 144, in solve
PyAEDT ERROR:     print(aedtapp.get_monitor_data())
PyAEDT ERROR:   File "C:\Users\nathm\venvs\pyaedt_232\lib\site-packages\pyaedt\application\Analysis.py", line 1904, in get_monitor_data
PyAEDT ERROR:     return self.desktop_class.get_monitor_data()
PyAEDT ERROR:   File "C:\Users\nathm\venvs\pyaedt_232\lib\site-packages\pyaedt\desktop.py", line 2280, in get_monitor_data
PyAEDT ERROR:     data = self.odesktop.GetMonitorData(reqstr)
PyAEDT ERROR:   File "C:\Users\nathm\venvs\pyaedt_232\lib\site-packages\pyaedt\generic\grpc_plugin.py", line 133, in DynamicFunc
PyAEDT ERROR:     return self.__Invoke__(funcName, args)
PyAEDT ERROR:   File "C:\Users\nathm\venvs\pyaedt_232\lib\site-packages\pyaedt\generic\grpc_plugin.py", line 124, in __Invoke__
PyAEDT ERROR:     raise GrpcApiError("Failed to execute grpc AEDT command: {}".format(funcName))
PyAEDT ERROR: AEDT grpc API call Error on get_monitor_data
PyAEDT ERROR: Last Electronics Desktop Message - [error] script macro error: can't get igenericmonitoringmgr (02:45:46 pm  jun 11, 2024)
```
For this test, I used AEDT 2023R2 and pyaedt 0.8.3

**Comment by Samuelopez-ansys** [comment]
@nathmay My bad, you are right, this is only for Icepak and monitors.

You can directly obtain the data if it is already solved:

```
import pyaedt

app = pyaedt.Hfss(specified_version="2024.1")

app.analyze(blocking=False)

data = app.post.get_solution_data()
y = data.data_real()
```
If it is not solved yet (if it is still doing the initial mesh), it will fail, so need to wait until you have a 1st adaptive mesh.

For that reason, I have discovered a nice way to get the profile info, that I will add in PyAEDT in the coming week, you can test it just doing:

from pyaedt.modeler.cad.elements3d import BinaryTreeNode

tree = BinaryTreeNode("profile", app.odesign.GetChildObject("results/profile"))

Tree contains all the Profile information, so you can get all the progress in real time :)

Best regards

**Reply by nathmay** [reply]
@Samuelopez-ansys Thanks for this, I tried this out and had some limited success. The returned BinaryTreeNode object sometimes lacked information, and sometimes it crashed the HFSS solver for some reason. I'll wait until you implement it, let me know if you need help in testing!

Another way I was exploring is to start analysis with `solve_in_batch=True`, as then AEDT records the profile in the `.batchinfo` folder. But then I came across a bug for `solve_in_batch`. Will file a bug report.

**Reply by Samuelopez-ansys** [reply]
@nathmay I have created this PR: https://github.com/ansys/pyaedt/pull/4799

You should wait until you have some data, because if the solver did not start, then the get profile raise an exception, but if you put it in a try-except, should work. But the method I have implemented already has this logic, so if the profile is empty, you will see an empty dictionary.

**Reply by JonathonMcMullan** [reply]
I was trying to do something similar (I want the tetrahedral count, DeltaS, and memory for the last adaptive pass). Is there a trick of extracting lists of strings from the BinaryTreeNode object that I have missed? 

![image](https://github.com/ansys/pyaedt/assets/162925770/8a4ae1d0-64f2-4afd-99b0-6ed24c1817a6)

**Reply by Samuelopez-ansys** [reply, marked as answer]
Hi @JonathonMcMullan,

Actually, the method get_profile is doing exactly this :)

https://aedt.docs.pyansys.com/version/stable/API/_autosummary/pyaedt.modules.SolveSetup.Setup.get_profile.html

If you check the code, you can also see how to get the information. 

tree.GetChildNames()

tree.GetChildObject(...)

**Reply by JonathonMcMullan** [reply]
Hi @Samuelopez-ansys,

Perfect, thanks very much. That did the trick.

![image](https://github.com/ansys/pyaedt/assets/162925770/45dd086e-d562-413d-a5d3-cb12ffcb1695)
